core/src/main/java/org/apache/iceberg/data/ContentFileWriteBuilder.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortOrder;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.apache.iceberg.io.DataWriter;
+
+/**
+ * A generic builder interface for creating specialized file writers in the Iceberg ecosystem.
+ *
+ * <p>This builder provides a unified configuration API for generating various types of content
+ * writers:
+ *
+ * <ul>
+ *   <li>{@link DataWriter} for creating data files with table records
+ *   <li>{@link EqualityDeleteWriter} for creating files with equality-based delete records
+ *   <li>{@link PositionDeleteWriter} for creating files with position-based delete records
+ * </ul>
+ *
+ * <p>Each concrete implementation configures the underlying file format writer while adding
+ * content-specific metadata and behaviors.
+ *
+ * @param <B> the concrete builder type for method chaining
+ */
+interface ContentFileWriteBuilder<B extends ContentFileWriteBuilder<B>> {
```

Comment:
I think there is probably a better name than `ContentFileWriteBuilder`. The term `ContentFile` is just a generic internal name that we use as a base for both `DataFile` and `DeleteFile` and it isn't a well-defined concept in the format. Leaking it into exposed APIs like this is not a good use. We are also going to be refactoring those classes for v4 metadata.

This interface is also a little strange. it looks like it has all of the methods from `WriteBuilder` except for `build`. It adds 4 methods to set metadata that get passed into `DataFile` and `DeleteFile` metadata: `spec`, `partition`, `key_metadata`, and `sortOrder`. I wonder if this could just be an interface with those last 4 that gets added to `DataWriteBuilder` and the delete builders instead?

Reply:
Also, the WriteBuilder API already has a `build` method which returns the appender, which we don't want to expose here

Reply:
The `ContentFileWriteBuilder` is not exposed, by I'm open for any suggestions 😄